### ISSUE TITLE
proc parameter in return type expression is not gensym'd inside templates

### ISSUE THREAD

**Issue by konsumlamm** [nim-lang/Nim#17433]
When using a (proc) parameter in the return type inside a template, it is not gensym'd.

### Example
```nim
from std/macros import expandMacros

expandMacros:
  template test() =
    proc foo(a: typedesc): a =
      default(a)
```

### Current Output
```
template main() =
  proc foo(a_385876008: typedesc): a =
    default(a_385876008)
```

### Expected Output
```
template main() =
  proc foo(a_385876008: typedesc): a_385876008 =
    default(a_385876008)
```

### Possible Solution
This can probably fixed in a similar fashion to #16874.

### Additional Information
This also happens when the parameter is used in a nested construct, like a function call.

I found this in #17432.

```
$ nim -v
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2021-03-21
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: b70e33f5bbb7b9c146c97d39aabae57504f4ee75
active boot switches: -d:release
```

EDIT: Use `typedesc`.

**Comment by timotheecour** [comment]
> I'm aware that the proc definition is not supposed to compile

can you replace by this, which is supposed to compile:

```nim
when defined case4: # works
  proc foo(a: typedesc): a = default(a)
  assert foo(float) == 0.0
  assert foo(string) == ""

when defined case5: # fails
  template main =
    proc foo(a: typedesc): a = default(a)
    assert foo(float) == 0.0
  main()

when defined case6:
  #[
template main() =
  proc foo(a_721420290: typedesc): a =
    default(a_721420290)

  assert foo(float) == 0.0
  ]#
  from std/macros import expandMacros
  expandMacros:
    template main =
      proc foo(a: typedesc): a = default(a)
      assert foo(float) == 0.0
```

**Comment by timotheecour** [comment]
> This can probably fixed in a similar fashion to #16874.

PR welcome!

**Comment by saem** [comment]
Hmm, I wonder if this needs a similar treatment as semProcAux recently got. I'll see if I can take a quick peek this weekend.
